libraries/AP_HAL/utility/RingBuffer.h

```diff
@@ -207,20 +208,150 @@ class ObjectBuffer {
+        delete buffer;
+    }
+
+    // Discards the buffer content, emptying it.
+    void clear(void) override
+    {
+        WITH_SEMAPHORE(sem);
+        ObjectBuffer<T>::clear();
+    }
+
+    // return number of objects available to be read from the front of the queue
+    uint32_t available(void) override
+    {
+        WITH_SEMAPHORE(sem);
+        return ObjectBuffer<T>::available();
+    }
+
+    // return number of objects that could be written to the back of the queue
+    uint32_t space(void) override
+    {
+        WITH_SEMAPHORE(sem);
+        return ObjectBuffer<T>::space();
+    }
+
+    // true is available() == 0
+    bool empty(void) override
+    {
+        WITH_SEMAPHORE(sem);
+        return ObjectBuffer<T>::empty();
+    }
+
+    // push one object onto the back of the queue
+    bool push(const T &object) override
+    {
+        WITH_SEMAPHORE(sem);
+        return ObjectBuffer<T>::push(object);
+    }
+
+    // push N objects onto the back of the queue
+    bool push(const T *object, uint32_t n) override
+    {
+        WITH_SEMAPHORE(sem);
+        return ObjectBuffer<T>::push(object, n);
+    }
+
+    /*
+      throw away an object from the front of the queue
+     */
+    bool pop(void) override
+    {
+        WITH_SEMAPHORE(sem);
+        return ObjectBuffer<T>::pop();
+    }
+
+    /*
+      pop earliest object off the front of the queue
+     */
+    bool pop(T &object) override
+    {
+        WITH_SEMAPHORE(sem);
+        return ObjectBuffer<T>::pop(object);
+    }
+
+    /*
+     * push_force() is semantically equivalent to:
+     *   if (!push(t)) { pop(); push(t); }
+     */
+    bool push_force(const T &object) override
+    {
+        WITH_SEMAPHORE(sem);
+        return ObjectBuffer<T>::push_force(object);
+    }
+
+    /*
+     * push_force() N objects
+     */
+    bool push_force(const T *object, uint32_t n) override
+    {
+        WITH_SEMAPHORE(sem);
+        return ObjectBuffer<T>::push_force(object, n);
+    }
+
+    /*
+      peek copies an object out from the front of the queue without advancing the read pointer
+     */
+    bool peek(T &object) override
+    {
+        WITH_SEMAPHORE(sem);
+        return ObjectBuffer<T>::peek(object);
+    }
+
+    /*
+      return a pointer to first contiguous array of available
+      objects. Return nullptr if none available
+     */
+    const T *readptr(uint32_t &n) override
+    {
+        WITH_SEMAPHORE(sem);
+        return ObjectBuffer<T>::readptr(n);
+    }
+
+    // advance the read pointer (discarding objects)
+    bool advance(uint32_t n) override
+    {
+        WITH_SEMAPHORE(sem);
+        return ObjectBuffer<T>::advance(n);
+    }
+
+    /* update the object at the front of the queue (the one that would
+       be fetched by pop()) */
+    bool update(const T &object) override
+    {
+        WITH_SEMAPHORE(sem);
+        return ObjectBuffer<T>::update(object);
+    }
+
+private:
+    ByteBuffer *buffer = nullptr;
+    HAL_Semaphore sem;
```

Comment:
this needs to be HAL_Semaphore_recursive

Reply:
fixed